### DISCUSSION TITLE
Create a "Home Assistant Content" system user to sign content

### DISCUSSION THREAD

**Discussion by balloob** [home-assistant/architecture#711] [General] [answered]
## Background

Content in Home Assistant requires authentication to be accessed. If we want to attach content to a notification or play it on a media player, the notification receiver and media player need to have authentication or they will not be able to access the content.

- For TTS we have implemented a hash in the URL.
- For camera we have a token that updates itself.
- Mobile apps will always access notifications that reference content in Home Assistant with authentication
- Cast will sign the paths to media source content with the "Home Assistant Cast" user (the user is used for the Cast application accessing HA)

## Problem

Notification receivers currently cannot embed camera snapshots via a URL unless they have and know about HA authentication. Ie, Discord.

Media players cannot play media hosted in Home Assistant (via media source) unless they have and know about HA authentication. Ie, Sonos.

The solution here is to [sign paths](https://github.com/home-assistant/core/blob/dev/homeassistant/components/http/auth.py#L32). A signed path takes a path and a user and created a signature that when added to the URL will grant the requestor access to that path in the name of that user.

```python
@callback
def async_sign_path(
    hass: HomeAssistant, refresh_token_id: str, path: str, expiration: timedelta
) -> str:
```

However signing paths requires a user and almost all integrations have no system generated user.

## Suggestion

Create a new system generated user "Home Assistant Content" and allow integrations to sign paths with this user.

Suggestion is to change the path to:

```python
@callback
def async_sign_path(
    hass: HomeAssistant, path: str, expiration: timedelta, *, refresh_token_id: str | None = None
) -> str:
```

This will automatically apply the Home Assistant Content refresh_token_id.

**Comment by pvizeli** [comment]
> For TTS we have implemented a hash in the URL

That is not for auth reason, we have to reference the speech in a way on the cache + on the API if you request something. We can sign that URL of a curse too. It's just to understand what we do that

**Comment by allenporter** [comment]
I have a couple questions:
- For those example use cases: Is it that its not *possible* to have a user associated with their requests or is it just complicated? I am not familiar with the context in which they are running.
- Would this replace the existing uses where paths are signed or just those that have a harder time understanding authentication?


Generally, my impression is that we'd strongly prefer to have authentication enforced at runtime and avoid use of signed urls whenever possible -- and so i'm wondering about what kinds of limitations would make sense to put in place for this, and also what guardrails to put in since its harder to remove use of signed urls once added.

Some thoughts:
- Can we limit the scope / permissions of what this user account is allowed to access -- e.g. only media and not other features in the API
- Set bounds on maximum expiration time for these types of use cases
- Use case specific registration to identify the source of requests or to allow ability to revoke 

(I'm biased from working on an effort to secure urls for a media serving system, but the scope of it is very different so some of these may not make sense for a more self contained system like home assistant)

**Reply by balloob** [reply]
So the main use case here is automations. 

If a user is available, we should use that authentication. We can even enforce that by using the `current_request` `ContextVar`.

**Comment by balloob** [comment, marked as answer]
Here is proposal for implementing this https://github.com/home-assistant/core/pull/64337
